icon.c

```diff
@@ -28,7 +37,144 @@ static double fit_to_square(int width, int height, int square_size) {
 	return longest > square_size ? square_size/longest : 1.0;
 }
 
-struct mako_icon *create_icon(const char *path, double max_size) {
+// Attempt to find a full path for a notification's icon_name, which may be:
+// - An absolute path, which will simply be returned (as a new string)
+// - A file:// URI, which will be converted to an absolute path
+// - A Freedesktop icon name, which will be resolved within the configured
+//   `icon-path` using something that looks vaguely like the algorithm defined
+//   in the icon theme spec (https://standards.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html)
+//
+// Returns the resolved path, or NULL if it was unable to find an icon. The
+// return value must be freed by the caller.
+static char *resolve_icon(struct mako_notification *notif) {
+	char *icon_name = notif->app_icon;
+	if (icon_name[0] == '\0') {
+		return NULL;
+	}
+	if (icon_name[0] == '/' || icon_name[0] == '~') {
```

Comment:
I don't think `~` will work. Probably better to leave it out for now.

Reply:
Good catch, I think I was originally planning to loop back and add an expansion of some sort and then forgot.